sdks/java/harness/src/main/java/org/apache/beam/fn/harness/status/BeamFnStatusClient.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.fn.harness.status;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.StringJoiner;
+import java.util.function.Function;
+import org.apache.beam.fn.harness.control.ProcessBundleHandler.BundleProcessor;
+import org.apache.beam.fn.harness.control.ProcessBundleHandler.BundleProcessorCache;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.WorkerStatusRequest;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.WorkerStatusResponse;
+import org.apache.beam.model.fnexecution.v1.BeamFnWorkerStatusGrpc;
+import org.apache.beam.model.pipeline.v1.Endpoints.ApiServiceDescriptor;
+import org.apache.beam.runners.core.metrics.ExecutionStateTracker;
+import org.apache.beam.vendor.grpc.v1p26p0.io.grpc.ManagedChannel;
+import org.apache.beam.vendor.grpc.v1p26p0.io.grpc.stub.StreamObserver;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BeamFnStatusClient {
+  private final StreamObserver<WorkerStatusResponse> outboundObserver;
+  private final BundleProcessorCache processBundleCache;
+  private static final Logger LOG = LoggerFactory.getLogger(BeamFnStatusClient.class);
+
+  public BeamFnStatusClient(
+      ApiServiceDescriptor apiServiceDescriptor,
+      Function<ApiServiceDescriptor, ManagedChannel> channelFactory,
+      BundleProcessorCache processBundleCache) {
+    BeamFnWorkerStatusGrpc.BeamFnWorkerStatusStub stub =
+        BeamFnWorkerStatusGrpc.newStub(channelFactory.apply(apiServiceDescriptor));
+    stub.workerStatus(new InboundObserver());
+    this.outboundObserver = stub.workerStatus(new InboundObserver());
+    this.processBundleCache = processBundleCache;
+  }
+
+  void threadDump(StringJoiner trace) {
```

Comment:
maybe call it getThreadDump? also could we do the deduplication like https://github.com/apache/beam/blob/master/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/status/ThreadzServlet.java#L76? Sometimes there are hundreds of threads and it is gonna make it hard for the browser to render.